txnsync/bitmask.go

```diff
@@ -0,0 +1,252 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"bytes"
+	"errors"
+)
+
+var errIndexNotFound = errors.New("invalid bitmask: index not found")
+var errInvalidBitmaskType = errors.New("invalid bitmask type")
+
+//msgp:allocbound bitmask maxBitmaskSize
+type bitmask []byte
+
+// assumed to be in mode 0, sets bit at index to 1
+func (b *bitmask) setBit(index int) {
+	byteIndex := index/8 + 1
+	(*b)[byteIndex] |= 1 << (index % 8)
+}
+
+// trimBitmask compresses the bitmask into one of the 4 types:
+// type 0: input bitmask bit pos x b -> output bitmask bit pos x b
+// type 1: input bitmask bit pos x b -> output bitmask bit pos x !b
+// type 2: stores the positions of bits where the bit value is 1
+//         input bitmask first bit 1 at pos A, second bit 1 at pos B, ...
+//         output bitmask stores A, B-A, ...
+// type 3: same as type 2, but stores the positons where the bit is 0
+func (b *bitmask) trimBitmask(entries int) {
+	if *b == nil {
+		return
+	}
+	lastExists := 0
+	lastNotExists := 0
+	numExists := 0
+	for i := 0; i < entries; i++ {
+		byteIndex := i/8 + 1
+		if (*b)[byteIndex]&(1<<(i%8)) != 0 {
+			lastExists = i
+			numExists++
+		} else {
+			lastNotExists = i
+		}
+	}
+	bitmaskType := 0
+	bestSize := bytesNeededBitmask(lastExists)
+	if bestSize > bytesNeededBitmask(lastNotExists) {
+		bitmaskType = 1
+		bestSize = bytesNeededBitmask(lastNotExists)
+	}
+	if bestSize > numExists*2+1 {
+		bitmaskType = 2
+		bestSize = numExists*2 + 1
+	}
+	if bestSize > (entries-numExists)*2+1 {
+		bitmaskType = 3
+		bestSize = (entries-numExists)*2 + 1
+	}
+	switch bitmaskType {
+	case 1:
```

Comment:
@nicholasguoalgorand and @algonautshant are familiar with this codebase. Could you take a look ?

Reply:
Here is an example for case 0.
*b = 00000000 00000001 11111111, entries = 8

With the current code (after fixing the issue with `bestSize`), `bestSize = 2` but the output will be same as original `*b`. Replacing `TrimRight()` with a slice resize should fix it though.

Reply:
This should not happen. the bitmask is created locally (not something received from the network), and locally it is not created with junk bits included in it which exceed `entries`. 
If you think there is a code path where this can happen, please describe it. It means you discovered a bug.

Reply:
Then I misunderstood what this function is supposed to do. Then why do we need the `entries` parameter? We could just scan the whole slice.